Issue #13465 - ensure that WebSocketClient onHandshakeResponse is not called twice

lachlan-roberts opened the pull request:
closes #13465

lachlan-roberts committed:
Issue #13465 - ensure that WebSocketClient onHandshakeResponse is not called twice

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts commented on jetty-core/jetty-websocket/jetty-websocket-core-client/src/main/java/org/eclipse/jetty/websocket/core/client/CoreClientUpgradeRequest.java:
```diff
@@ -261,14 +263,21 @@ public void onComplete(Result result)
-                "Failed to upgrade to websocket: Unexpected HTTP Response Status Code: " + responseLine));
+
+            if (result.getFailure() != null)
+                handleException(new UpgradeException(requestURI, status, result.getFailure().getMessage()));
```
I think this also fixes https://github.com/jetty/jetty.project/issues/13466

lachlan-roberts replied:
This change seems to be resulting in a lot of flakiness in the HTTP/2 WebSocket tests. 

> "Unexpected HTTP Response Status Code: 500" ignoring case
     but: was "org.eclipse.jetty.websocket.core.exception.UpgradeException: Stream has been reset"

sbordet reviewed (changes requested):
@lachlan-roberts yes it should also fix #13466.

The fix should be less flaky, so the test failures should be investigated.

lachlan-roberts commented:
@sbordet I have investigated the failures.

In the test the client is trying to upgrade to websocket over HTTP/2 and the server is throwing an exception in the websocket creator.

When the client sends the headers frame to upgrade it does not include an END_STREAM flag because the subsequent data frames it sends will be websocket bytes.

The server sees the exception and tries to send a 500 response, it sends a HEADERS frame with the END_STREAM flag.
But the server sees that the client has not sent an END_STREAM flag so it resets the stream.

The client then has a race whether it processes the HEADERS frame with the 500 response, or the RST_STREAM frame first.

sbordet committed:
Fixed failing tests.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
@lachlan-roberts I fixed the tests, please review the fix.

The build failed due to #13533, so I think this PR is fine now.

lachlan-roberts committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into jetty-12.1.x-13465-WebSocketClientResponseListener

lachlan-roberts commented:
@sbordet it seems after merging with `12.1.x` it still has the same failures.
CI shows the `Stream has been reset` failure is still there in `WebSocketOverHTTP2Test`.

sbordet committed:
Updated handling of HttpStreamOverHTTP2 success and failure.

Now testing whether the request was a tunnel, and if so, try to close the request side of the stream.
If the close is successful, it means a response was already sent, so we can avoid sending a RST_STREAM.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Made sure the HTTP/2 client closes the request side in case of failed tunnel.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
